packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -441,18 +476,83 @@ export class LoaderContainerTracker implements IOpProcessingController {
 	 * Any containers given that is not tracked will be ignored.
 	 */
 	public async pauseProcessing(...containers: IContainer[]) {
-		const pauseP: Promise<void>[] = [];
+		const waitP: Promise<void>[] = [];
 		const containersToApply = this.getContainers(containers);
 		for (const container of containersToApply) {
 			const record = this.containers.get(container);
 			if (record !== undefined && !record.paused) {
-				debugWait(`${record.index}: container paused`);
-				pauseP.push(container.deltaManager.inbound.pause());
-				pauseP.push(container.deltaManager.outbound.pause());
-				record.paused = true;
+				if (record.pauseP === undefined) {
+					record.pauseP = this.pauseContainer(container, record);
+				}
+				waitP.push(record.pauseP);
 			}
 		}
-		await Promise.all(pauseP);
+		await Promise.all(waitP);
+	}
+
+	/**
+	 * When we pause a container, it is assumed that we want fine grain control over
+	 * op sequencing.  This function will prepare the container so that it will be
+	 * connected in write mode for that purpose (to avoid missing join messages or
+	 * change the sequence of event when switching from read to write mode)
+	 *
+	 * @param container - the container to pause
+	 * @param record - the record for the container
+	 */
+	private async pauseContainer(container: IContainer, record: ContainerRecord) {
+		debugWait(`${record.index}: pausing container`);
+		assert(!container.deltaManager.outbound.paused, "Container should not be paused yet");
+		assert(!container.deltaManager.inbound.paused, "Container should not be paused yet");
+
+		// Pause outbound
+		debugWait(`${record.index}: pausing container outbound queues`);
+		await container.deltaManager.outbound.pause();
+
+		// Ensure the container is connected first.
+		if (container.connectionState !== ConnectionState.Connected) {
+			debugWait(`${record.index}: Wait for container connection`);
+			await waitForContainerConnection(container);
+		}
+
+		// Check if the container is in write mode
+		if (!container.deltaManager.active) {
```

Comment:
I don't follow what is happening inside this if block. Does a code proposal switch the container to write mode? Is that what this code is doing?

Reply:
If so, should we just assert that the container is in write mode when this or one of the process methods is called? And that becomes the contract of these APIs. It's up to the caller to ensure that the container is in write mode if they use these APIs rather than making an implicit decision to switch container to write mode.

Reply:
It's simple for tests to transition a container to write mode. A lof of our tests already do this.

Reply:
As per our discussion, cam you please add comments on the APIs that would result in switch to write mode. Basically, users of the API should know that the container will switch to write mode if it's in read mode.